README
Scala implementation of GoGrid client.



Changelog
v1.1
* list of servers added
* list of jobs added


src/main/scala/info/vitko/gogrid/Image.scala
package info.vitko.gogrid

import argonaut.Argonaut._
import argonaut.CodecJson

/**
 * Created with IntelliJ IDEA.
 * User: vist
 * Date: 22/09/14
 * Time: 21:57
 */
case class Image(isActive: Boolean,
                 updatedTime: Int,
                 isPublic: Boolean,
                 name: String,
                 imageType: GeneralOption,
                 price: Int,
                 obj: String,
                 owner: ImageOwner,
                 datacenterlist: List[ImageDatacenter],
                 state: GeneralOption,
                 friendlyName: String,
                 gsitype: GeneralOption,
                 location: String,
                 createdTime: Option[Int],
                 minram: GeneralOption,
                 os: GeneralOption,
                 id: Int,
                 billingtokens: List[BillingToken],
                 architecture: GeneralOption)

object Image {
  implicit def ImageCodecJson: CodecJson[Image] =
    casecodec19(Image.apply, Image.unapply)("isActive", "updatedTime", "isPublic", "name", "type", "price",
      "object", "owner", "datacenterlist", "state", "friendlyName", "gsitype", "location", "createdTime", "minram",
      "os", "id", "billingtokens", "architecture")
}
case class ImageOwner(obj: String, id: Int, name: String)

object ImageOwner{
  implicit def ImageOwnerCodecJson: CodecJson[ImageOwner] =
    casecodec3(ImageOwner.apply, ImageOwner.unapply)("object", "id", "name")
}

case class ImageDatacenter(createdTime: Option[Int], datacenter: GeneralOption, state: GeneralOption,
                           obj: String, updatedTime: Int)

object ImageDatacenter {
  implicit def ImageDatacenterCodecJson: CodecJson[ImageDatacenter] =
    casecodec5(ImageDatacenter.apply, ImageDatacenter.unapply)("createdTime", "datacenter", "state", "object",
      "updatedTime")
}

case class BillingToken(price: Int, obj: String, id: Int, name: String)

object BillingToken{
  implicit def BillingTokenCodecJson: CodecJson[BillingToken] =
    casecodec4(BillingToken.apply, BillingToken.unapply)("price", "object", "id", "name")
}


src/main/scala/info/vitko/gogrid/Server.scala
package info.vitko.gogrid

import argonaut.Argonaut._
import argonaut.CodecJson

/**
 * Created with IntelliJ IDEA.
 * User: vist
 * Date: 22/09/14
 * Time: 21:42
 */

case class Server(datacenter: GeneralOption,
                  ram: GeneralOption,
                  description: Option[String],
                  srvType: GeneralOption,
                  image: Image,
                  obj: String,
                  state: GeneralOption,
                  isSandbox: Boolean,
                  ip: ServerIP,
                  cores: Option[Int],
                  diskSize: Option[Int],
                  os: GeneralOption,
                  id: Option[Int],
                  name: String) {

  override def toString() = s"${id.getOrElse(None)}\t$name\t$ip"
}

object Server {
  implicit def ServerCodecJson: CodecJson[Server] =
    casecodec14(Server.apply, Server.unapply)("datacenter", "ram", "description", "type", "image", "object", "state",
      "isSandbox", "ip", "cores", "diskSize", "os", "id", "name")
}

case class ServersResponse(servers: List[Server], method: String, status: String, summary: GoGridResponseSummary)

object ServersResponse {
  implicit def ServersResponseCodecJson: CodecJson[ServersResponse] =
    casecodec4(ServersResponse.apply, ServersResponse.unapply)("list", "method", "status", "summary")
}

case class ServerIP(subnet: String,
                    datacenter: GeneralOption,
                    ip: String,
                    obj: String,
                    public: Boolean,
                    state: GeneralOption,
                    id: Int) {

  override def toString() = ip
}

object ServerIP {
  implicit def ServerIPCodecJson: CodecJson[ServerIP] =
    casecodec7(ServerIP.apply, ServerIP.unapply)("subnet", "datacenter", "ip", "object", "public", "state","id")
}

src/main/scala/info/vitko/gogrid/GoGridClient.scala
package info.vitko.gogrid

import java.security.MessageDigest

import dispatch.Defaults._
import dispatch._

import scalaz._, Scalaz._
import argonaut._, Argonaut._
/**
 * Created with IntelliJ IDEA.
 * User: vist
 * Date: 17/09/14
 * Time: 17:03
 */
class GoGridClient(key: String, secret: String,
                   apiUrl: String = GoGridConfig.apiUrl,
                   version: String = GoGridConfig.version) {

  def this() = this(GoGridConfig.AccessConfig.key, GoGridConfig.AccessConfig.secret)

  def listServers() = {
    val response = Parse.decodeOption[ServersResponse](apiRequest("grid/server/list"))
    response.get.servers
  }

  def listJobs() = {
    val decoded = Parse.decodeOption[JobsResponse](apiRequest("grid/job/list"))
    decoded.get.jobs
  }

  private def apiRequest(path: String, format: String = "json") = {
    val requestUrl = s"${this.apiUrl}/${path}"
    val svc = url(requestUrl) <<? Map("api_key" -> key) <<? Map("sig" -> getSig()) <<?
      Map("v" -> version) <<? Map("format" -> format)
//    println(svc.toRequest)
    val respF = Http(svc OK as.String)

    // add handler to shutdown NIO engine on complete
//    respF onComplete { case _ => Http.shutdown() }

    respF()
  }

  private def getSig() = {
    val timestamp = System.currentTimeMillis / 1000
    val string = key + secret + timestamp
    val digest = MessageDigest.getInstance("MD5")
    digest.digest(string.getBytes).map("%02x".format(_)).mkString
  }

}

object GoGridClient extends App {
  // create GoGridClient
  val client = new GoGridClient()

  // list of jobs
  val jbs = client.listJobs()
  for (job <- jbs) println(job)

  // list of servers
  for (srv <- client.listServers()) println(srv)

  Http.shutdown()
}


src/main/scala/info/vitko/gogrid/GoGridJob.scala
package info.vitko.gogrid

import scalaz._, Scalaz._
import argonaut._, Argonaut._

/**
 * Created with IntelliJ IDEA.
 * User: vist
 * Date: 19/09/14
 * Time: 16:01
 */

case class GoGridJob(attempts: Int,
                     command: GeneralOption,
                     createdon: Int,
                     currentstate: GeneralOption,
                     datacenter: GeneralOption,
                     detail: JobDetail,
                     history: List[JobHistoryEntry],
                     id: Int,
                     lastupdatedon: Option[Int],
                     obj: String,
                     objecttype: GeneralOption,
                     owner: String) {

  override def toString = s"$createdon by $owner\t$command\t$detail"
}

case class JobDetail(description: Option[String], image: String, ip: String, name: String, detailType: String){
  override def toString() = s"$ip\t$name"
}

object JobDetail {
  implicit def JobDetailCodecJson: CodecJson[JobDetail] =
    casecodec5(JobDetail.apply, JobDetail.unapply)("description", "image", "ip", "name", "type")
}

case class JobHistoryEntry(id: Int, state: GeneralOption, updatedon: Option[Int])

object JobHistoryEntry {
  implicit def JobHistoryEntryCodecJson: CodecJson[JobHistoryEntry] =
    casecodec3(JobHistoryEntry.apply, JobHistoryEntry.unapply)("id", "state", "updatedon")
}

object GoGridJob {
  implicit def GoGridJobCodecJson: CodecJson[GoGridJob] =
    casecodec12(GoGridJob.apply, GoGridJob.unapply)("attempts", "command", "createdon", "currentstate",
      "datacenter", "detail", "history", "id", "lastupdatedon", "object", "objecttype", "owner")
}

case class JobsResponse(jobs: List[GoGridJob], method: String, status: String, summary: GoGridResponseSummary)

object JobsResponse {
  implicit def JobsResponseCodecJson: CodecJson[JobsResponse] =
    casecodec4(JobsResponse.apply, JobsResponse.unapply)("list", "method", "status", "summary")
}